### DISCUSSION TITLE
Is it possible to add metadata before text is processed by the Pipeline

### DISCUSSION THREAD

**Discussion by markFriel** [explosion/spaCy#8204] [Help: Coding & Implementations] [answered]
Hi,

I'm not sure if this is possible as haven't seen anything around it.

I have a case where I have the raw text and a  list equal to the number of tokens in the raw_text that represent where each token came from in an text. I am doing NER and I would like to create a custom component in which I can treat the entities that are extracted from the NER component differently depending on where the tokens have come from in the text

```
text = 'This is an example text! \n This is example text'
token_metdata = ['line_1','line_1','line_1','line_1','line_1','line_1','line_1','line_2',line_2',line_2',line_2',]

#Example of adding metadata to the pipeline 
doc = nlp(text, token_metadata)
```

I have seen the documentation surrounding extensions, but I believe the text will already have been processed at that stage, whereas I need the metadata before the text has been passed through the custom component.

Any help would be appreciated on how to achieve this.

Thanks

**Comment by polm** [comment, marked as answer]
Actually had question about the same thing recently here.

https://github.com/explosion/spaCy/discussions/8194

Basically there's not a default way to do this, but there are a couple of ways to work around it. See that thread for details, but let me know if you have any questions about it.

**Reply by polm** [reply]
Ah, also note that because your data is per-token, it may have issues if your tokenization differs from spaCy's.

**Reply by markFriel** [reply]
Hi,

Yes,  it looks to be a similar use case, thank you for pointing me towards it. And yes your 100% correct I had to generate the token level metadata by using the spacy tokenizer. 

Thanks for getting back to me.
